From fa6131e7e7562aec37845112ed397fe77f67ed38 Mon Sep 17 00:00:00 2001
From: Arun Ravindran <arun.ravindran@intel.com>
Date: Wed, 21 Nov 2012 17:29:05 +0200
Subject: [PATCH 2/2] [PORT FROM MAIN] ril: Enabling seek and Smart Card API
 support in JB 4.2

BZ: 47372

The following changes are ported from MCG Android 4.1.2
for enabling telephony in 4.2.

1) Introduction of new RIL request for support of
   smart card API
2) Introduction of new RIL_Errno enumerations for
   seek and smart card API.

Change-Id: I1a06f5bce7eb5a1c4b3c4316553b8bfe5a321afc
Signed-off-by: Arun Ravindran <arun.ravindran@intel.com>
Reviewed-on: http://android.intel.com:8080/75274
Reviewed-by: Fremaux, Pascal <pascal.fremaux@intel.com>
Reviewed-by: Badrappan, Jeevaka <jeevaka.badrappan@intel.com>
Reviewed-by: Yao, Yong <yong.yao@intel.com>
Reviewed-by: Lebsir, SamiX <samix.lebsir@intel.com>
Tested-by: Lebsir, SamiX <samix.lebsir@intel.com>
Reviewed-by: Protat, Remy <remy.protat@intel.com>
Tested-by: Protat, Remy <remy.protat@intel.com>
---
 include/telephony/ril.h |   20 ++++++++++++++++++++
 libril/ril.cpp          |   11 +++++++++++
 libril/ril_commands.h   |    4 ++++
 3 files changed, 35 insertions(+), 0 deletions(-)

diff --git a/include/telephony/ril.h b/include/telephony/ril.h
index c7be414..a7c61cc 100644
--- a/include/telephony/ril.h
+++ b/include/telephony/ril.h
@@ -63,6 +63,8 @@ typedef enum {
     RIL_E_MISSING_RESOURCE = 17,
     RIL_E_NO_SUCH_ELEMENT = 18,
 
+    RIL_E_INVALID_PARAMETER = 19                /* SEEK for Android */
+
 } RIL_Errno;
 
 typedef enum {
@@ -273,6 +275,7 @@ typedef struct {
 } RIL_Dial;
 
 typedef struct {
+    int cla;
     int command;    /* one of the commands listed for TS 27.007 +CRSM*/
     int fileid;     /* EF id */
     char *path;     /* "pathid" from TS 27.007 +CRSM command.
@@ -287,6 +290,7 @@ typedef struct {
 } RIL_SIM_IO_v5;
 
 typedef struct {
+    int cla;
     int command;    /* one of the commands listed for TS 27.007 +CRSM*/
     int fileid;     /* EF id */
     char *path;     /* "pathid" from TS 27.007 +CRSM command.
@@ -3328,6 +3332,22 @@ typedef struct {
  */
 #define RIL_REQUEST_VOICE_RADIO_TECH 108
 
+// "data" is a const RIL_SIM_IO *
+// "response" is a const RIL_SIM_IO_Response *
+#define RIL_REQUEST_SIM_TRANSMIT_BASIC 109
+
+// "data" is a const char * containing the AID of the applet
+// "response" is a int * containing the channel id
+#define RIL_REQUEST_SIM_OPEN_CHANNEL 110
+
+// "data" is a const int * containing the channel id
+// "response" is NULL
+#define RIL_REQUEST_SIM_CLOSE_CHANNEL 111
+
+// "data" is a const RIL_SIM_IO *
+// "response" is a const RIL_SIM_IO_Response *
+#define RIL_REQUEST_SIM_TRANSMIT_CHANNEL 112
+
 /***********************************************************************/
 
 
diff --git a/libril/ril.cpp b/libril/ril.cpp
index 633b3dc..86115bb 100644
--- a/libril/ril.cpp
+++ b/libril/ril.cpp
@@ -689,6 +689,7 @@ invalid:
 /**
  * Callee expects const RIL_SIM_IO *
  * Payload is:
+ *   int32_t cla
  *   int32_t command
  *   int32_t fileid
  *   String path
@@ -712,6 +713,12 @@ dispatchSIM_IO (Parcel &p, RequestInfo *pRI) {
 
     // note we only check status at the end
 
+    simIO.v6.cla = 0;
+    if (pRI->pCI->requestNumber != RIL_REQUEST_SIM_IO) {
+        status = p.readInt32(&t);
+        simIO.v6.cla = (int)t;
+    }
+
     status = p.readInt32(&t);
     simIO.v6.command = (int)t;
 
@@ -3371,6 +3378,10 @@ requestToString(int request) {
         case RIL_REQUEST_SEND_SMS_EXPECT_MORE: return "SEND_SMS_EXPECT_MORE";
         case RIL_REQUEST_SETUP_DATA_CALL: return "SETUP_DATA_CALL";
         case RIL_REQUEST_SIM_IO: return "SIM_IO";
+        case RIL_REQUEST_SIM_TRANSMIT_BASIC: return "SIM_TRANSMIT_BASIC";
+        case RIL_REQUEST_SIM_OPEN_CHANNEL: return "SIM_OPEN_CHANNEL";
+        case RIL_REQUEST_SIM_CLOSE_CHANNEL: return "SIM_CLOSE_CHANNEL";
+        case RIL_REQUEST_SIM_TRANSMIT_CHANNEL: return "SIM_TRANSMIT_CHANNEL";
         case RIL_REQUEST_SEND_USSD: return "SEND_USSD";
         case RIL_REQUEST_CANCEL_USSD: return "CANCEL_USSD";
         case RIL_REQUEST_GET_CLIR: return "GET_CLIR";
diff --git a/libril/ril_commands.h b/libril/ril_commands.h
index 069ee75..45d309b 100644
--- a/libril/ril_commands.h
+++ b/libril/ril_commands.h
@@ -123,3 +123,7 @@
     {RIL_REQUEST_ACKNOWLEDGE_INCOMING_GSM_SMS_WITH_PDU, dispatchStrings, responseVoid},
     {RIL_REQUEST_STK_SEND_ENVELOPE_WITH_STATUS, dispatchString, responseSIM_IO},
     {RIL_REQUEST_VOICE_RADIO_TECH, dispatchVoiceRadioTech, responseInts},
+    {RIL_REQUEST_SIM_TRANSMIT_BASIC, dispatchSIM_IO, responseSIM_IO},
+    {RIL_REQUEST_SIM_OPEN_CHANNEL, dispatchString, responseInts},
+    {RIL_REQUEST_SIM_CLOSE_CHANNEL, dispatchInts, responseVoid},
+    {RIL_REQUEST_SIM_TRANSMIT_CHANNEL, dispatchSIM_IO, responseSIM_IO}
-- 
1.7.9

